test/reverse_gpu.jl

```diff
@@ -0,0 +1,63 @@
+using Test
+using Enzyme
+using KernelAbstractions
+using CUDA
+
+@kernel function square!(A)
+    I = @index(Global, Linear)
+    @inbounds A[I] *= A[I]
+end
+
+function square_caller(A, backend)
+    kernel = square!(backend)
+    kernel(A, ndrange=size(A))
+    KernelAbstractions.synchronize(backend)
+end
+
+
+@kernel function mul!(A, B)
+    I = @index(Global, Linear)
+    @inbounds A[I] *= B
+end
+
+function mul_caller(A, B, backend)
+    kernel = mul!(backend)
+    kernel(A, B, ndrange=size(A))
+    KernelAbstractions.synchronize(backend)
```

Comment:
@vchuravy is there a KA way to have the kernel run immediately rather than get sync'd. That way this can be separately merged from adding support for KA async

Reply:
I am not sure what you mean by that? CUDA etc are all async, so the synchronized is needed in any case. You can leave it out and rely on stream ordering?

Reply:
Ok in which case we need to add KA sync rules for forward and reverse. Forward is easy, just sync the regular and shadow kernels. Reverse needs to launch the reverse kernel. and reverse kernel execute needs to sync it